File: src/main/java/com/shopjoy/service/ProductService.java
package com.shopjoy.service;

import com.shopjoy.dto.filter.ProductFilter;
import com.shopjoy.dto.request.CreateProductRequest;
import com.shopjoy.dto.request.UpdateProductRequest;
import com.shopjoy.dto.response.ProductResponse;
import com.shopjoy.entity.Product;
import com.shopjoy.util.Page;
import com.shopjoy.util.Pageable;

import java.util.List;

/**
 * The interface Product service.
 */
public interface ProductService {

    /**
     * Create product product response.
     *
     * @param request the request
     * @return the product response
     */
    ProductResponse createProduct(CreateProductRequest request);

    /**
     * Gets product by id.
     *
     * @param productId the product id
     * @return the product by id
     */
    ProductResponse getProductById(Integer productId);

    /**
     * Gets all products.
     *
     * @return the all products
     */
    List<ProductResponse> getAllProducts();

    /**
     * Gets active products.
     *
     * @return the active products
     */
    List<ProductResponse> getActiveProducts();

    /**
     * Gets products by category.
     *
     * @param categoryId the category id
     * @return the products by category
     */
    List<ProductResponse> getProductsByCategory(Integer categoryId);

    /**
     * Search products by name list.
     *
     * @param keyword the keyword
     * @return the list
     */
    List<ProductResponse> searchProductsByName(String keyword);

    /**
     * Gets products by price range.
     *
     * @param minPrice the min price
     * @param maxPrice the max price
     * @return the products by price range
     */
    List<ProductResponse> getProductsByPriceRange(double minPrice, double maxPrice);

    /**
     * Update product product response.
     *
     * @param productId the product id
     * @param request   the request
     * @return the product response
     */
    ProductResponse updateProduct(Integer productId, UpdateProductRequest request);

    /**
     * Update product price product response.
     *
     * @param productId the product id
     * @param newPrice  the new price
     * @return the product response
     */
    ProductResponse updateProductPrice(Integer productId, double newPrice);

    /**
     * Activate product product response.
     *
     * @param productId the product id
     * @return the product response
     */
    ProductResponse activateProduct(Integer productId);

    /**
     * Deactivate product product response.
     *
     * @param productId the product id
     * @return the product response
     */
    ProductResponse deactivateProduct(Integer productId);

    /**
     * Delete product.
     *
     * @param productId the product id
     */
    void deleteProduct(Integer productId);

    /**
     * Gets total product count.
     *
     * @return the total product count
     */
    long getTotalProductCount();

    /**
     * Gets product count by category.
     *
     * @param categoryId the category id
     * @return the product count by category
     */
    long getProductCountByCategory(Integer categoryId);

    /**
     * Gets products paginated.
     *
     * @param pageable      the pageable
     * @param sortBy        the sort by
     * @param sortDirection the sort direction
     * @return the products paginated
     */
    Page<ProductResponse> getProductsPaginated(Pageable pageable, String sortBy, String sortDirection);

    /**
     * Search products paginated page.
     *
     * @param searchTerm the search term
     * @param pageable   the pageable
     * @return the page
     */
    Page<ProductResponse> searchProductsPaginated(String searchTerm, Pageable pageable);

    /**
     * Gets products with filters.
     *
     * @param filter        the filter
     * @param pageable      the pageable
     * @param sortBy        the sort by
     * @param sortDirection the sort direction
     * @return the products with filters
     */
    Page<ProductResponse> getProductsWithFilters(ProductFilter filter, Pageable pageable, String sortBy,
            String sortDirection);

    /**
     * Gets products sorted with quick sort.
     *
     * @param sortBy    the sort by
     * @param ascending the ascending
     * @return the products sorted with quick sort
     */
    List<ProductResponse> getProductsSortedWithQuickSort(String sortBy, boolean ascending);

    /**
     * Gets products sorted with merge sort.
     *
     * @param sortBy    the sort by
     * @param ascending the ascending
     * @return the products sorted with merge sort
     */
    List<ProductResponse> getProductsSortedWithMergeSort(String sortBy, boolean ascending);

    /**
     * Search product by id with binary search product response.
     *
     * @param productId the product id
     * @return the product response
     */
    ProductResponse searchProductByIdWithBinarySearch(Integer productId);

    List<ProductResponse> findAllSorted(String sortBy, String sortDirection, String algorithm);

    Product searchById(Integer id);

    List<ProductResponse> getRecentlyAddedProducts(int limit);
}

File: frontend/src/services/productService.js
import api from './api';

// POST /api/v1/products - Create product
export const createProduct = (productData) => api.post('/products', productData);

// GET /api/v1/products/{id} - Get product by ID
export const getProductById = (id) => api.get(`/products/${id}`);

// GET /api/v1/products - Get all products
export const getAllProducts = () => api.get('/products');

// GET /api/v1/products/active - Get active products
export const getActiveProducts = () => api.get('/products/active');

// GET /api/v1/products/category/{categoryId} - Get products by category
export const getProductsByCategory = (categoryId) => api.get(`/products/category/${categoryId}`);

// GET /api/v1/products/search?name={name} - Search products by name
export const searchProductsByName = (name) => api.get('/products/search', { params: { name } });

// GET /api/v1/products/price-range?minPrice={min}&maxPrice={max} - Get products by price range
export const getProductsByPriceRange = (minPrice, maxPrice) =>
    api.get('/products/price-range', { params: { minPrice, maxPrice } });

// PUT /api/v1/products/{id} - Update product
export const updateProduct = (id, productData) => api.put(`/products/${id}`, productData);

// PATCH /api/v1/products/{id}/price?newPrice={price} - Update product price
export const updateProductPrice = (id, newPrice) =>
    api.patch(`/products/${id}/price`, null, { params: { newPrice } });

// PATCH /api/v1/products/{id}/activate - Activate product
export const activateProduct = (id) => api.patch(`/products/${id}/activate`);

// PATCH /api/v1/products/{id}/deactivate - Deactivate product
export const deactivateProduct = (id) => api.patch(`/products/${id}/deactivate`);

// DELETE /api/v1/products/{id} - Delete product
export const deleteProduct = (id) => api.delete(`/products/${id}`);

// GET /api/v1/products/count - Get total product count
export const getTotalProductCount = () => api.get('/products/count');

// GET /api/v1/products/count/category/{categoryId} - Get product count by category
export const getProductCountByCategory = (categoryId) => api.get(`/products/count/category/${categoryId}`);

// GET /api/v1/products/paginated - Get products with pagination
export const getProductsPaginated = (page = 0, size = 10, sortBy = 'product_id', sortDirection = 'ASC') =>
    api.get('/products/paginated', { params: { page, size, sortBy, sortDirection } });

// GET /api/v1/products/search/paginated - Search products with pagination
export const searchProductsPaginated = (term, page = 0, size = 10) =>
    api.get('/products/search/paginated', { params: { term, page, size } });

// GET /api/v1/products/filter - Get products with filters
export const getProductsWithFilters = (filters) =>
    api.get('/products/filter', { params: filters });

// GET /api/v1/products/sorted/quicksort - Get products sorted with QuickSort
export const getProductsSortedWithQuickSort = (sortBy = 'product_id', ascending = true) =>
    api.get('/products/sorted/quicksort', { params: { sortBy, ascending } });

// GET /api/v1/products/sorted/mergesort - Get products sorted with MergeSort
export const getProductsSortedWithMergeSort = (sortBy = 'product_id', ascending = true) =>
    api.get('/products/sorted/mergesort', { params: { sortBy, ascending } });

// GET /api/v1/products/{id}/binary-search - Search product by ID with Binary Search
export const searchProductByIdWithBinarySearch = (id) => api.get(`/products/${id}/binary-search`);

// GET /api/v1/products/algorithms/sort-comparison - Compare sorting algorithms
export const compareSortingAlgorithms = (datasetSize = 1000) =>
    api.get('/products/algorithms/sort-comparison', { params: { datasetSize } });

// GET /api/v1/products/algorithms/search-comparison - Compare search algorithms
export const compareSearchAlgorithms = (datasetSize = 1000) =>
    api.get('/products/algorithms/search-comparison', { params: { datasetSize } });

// GET /api/v1/products/algorithms/recommendations - Get algorithm recommendations
export const getAlgorithmRecommendations = (datasetSize = 5000) =>
    api.get('/products/algorithms/recommendations', { params: { datasetSize } });

// GET /api/v1/products/sorted/{algorithm} - Get products with custom sorting algorithm
export const getProductsWithAlgorithm = (algorithm, sortBy = 'price', sortDirection = 'ASC') =>
    api.get(`/products/sorted/${algorithm}`, { params: { sortBy, sortDirection } });

// GET /api/v1/products/new-arrivals - Get recently added products
export const getNewArrivals = (limit = 10) =>
    api.get('/products/new-arrivals', { params: { limit } });
